### PR TITLE
gh-121267: Improve performance of tarfile (#121267)

### PR DESCRIPTION
**Pull request by jforberg**
Tarfile in the default write mode spends much of its time resolving UIDs into usernames and GIDs into group names. By caching these mappings, a significant speedup can be achieved.

In my simple benchmark[1], this extra caching speeds up tarfile by 8x.

[1] https://gist.github.com/jforberg/86af759c796199740c31547ae828aef2

* Issue: gh-121267

### CODE REVIEW

**Comment by nineteendo** [comment]
Can the cache get outdated during execution?

**Comment by jforberg** [comment]
@nineteendo Yes. Suppose the passwd database changes during processing of a tar file. We have two options:

1. Generate a tar file with inconsistent UID->uname mapping. Files owned by the same user will have different uname.
2. Keep the same uname as when the UID was first encountered, ignore the new change.

Python currently does (1) which comes at a steep (several hundred percent) performance cost as we need to re-read and parse the passwd/group database for every single file. Note also that we first stat the file, *then* read passwd so there is already a race condition in the code if we want to view it that way.
 
Doing (2) is much faster as I've shown. It's also what GNU tar does, so it's not a new idea.

**Comment by gaogaotiantian** [comment]
> Generate a tar file with inconsistent UID->uname mapping. Files owned by the same user will have different uname.

But you are doing the cache at class level right? Which means it might not be a "single" file that has an inconsistent mapping. If it's a long-running process, the user could generate a tar file, then changed the mapping, and try to generate another one - but the name is cached and the change won't be reflected.

**Comment by nineteendo** [comment]
Yeah, that's what I was thinking too, would a new parameter be better?

```diff
 def __init__(self, name=None, mode="r", fileobj=None, format=None,
         tarinfo=None, dereference=None, ignore_zeros=None, encoding=None,
         errors="surrogateescape", pax_headers=None, debug=None,
-        errorlevel=None, copybufsize=None, stream=False):
+        errorlevel=None, copybufsize=None, stream=False, uname_cache=None,
+        gname_cache=None):
```

**Comment by jforberg** [comment]
> But you are doing the cache at class level right?

I'm sorry, that was a mistake. I intended the cache to be per TarFile instance, to make the lifetime of the cache clearly limited. I'd be happy to correct it. I agree that a process-global cache isn't a good idea.

@nineteendo I'm not sure I follow you. Do you mean to make the cachin opt-in? I was hoping that this could be turned on by default, so more code can benefit from the speedup.

**Comment by nineteendo** [comment]
Nevermind, just put it on the instance (there's already a cache for inodes):

https://github.com/python/cpython/blob/41397cee70f4933b539ab1a9232f04419ac8e2bb/Lib/tarfile.py#L1724-L1732

**Comment by jforberg** [comment]
@nineteendo @gaogaotiantian Thanks for your feedback! I have pushed a fixed version now.

**Comment by gaogaotiantian** [comment]
Next time please do not force push - it would be easier for us to review the code and the history.

File: Lib/tarfile.py
Comment (gaogaotiantian) [inline Lib/tarfile.py]:
```suggestion
            if tarinfo.uid not in self.uname_cache:
                try:
                    self.uname_cache[tarinfo.uid] = pwd.getpwuid(tarinfo.uid)[0]
                except KeyError:
                    self.uname_cache[tarinfo.uid] = None
            tarinfo.uname = self.uname_cache[tarinfo.uid]
```

File: Lib/tarfile.py
Comment (gaogaotiantian) [inline Lib/tarfile.py]:
```suggestion
            if tarinfo.gid not in self.gname_cache:
                try:
                    self.gname_cache[tarinfo.gid] = grp.getgrgid(tarinfo.gid)[0]
                except KeyError:
                    self.gname_cache[tarinfo.gid] = None
            tarinfo.gname = self.gname_cache[tarinfo.gid]
```

**Review by gaogaotiantian — CHANGES_REQUESTED** [review]
`a not in b` is preferred than `not a in b` as it's easier to read. Also if `pwd.getpwuid` raises a `KeyError`, your current code does not cache it. I made some suggestions, feel free to just apply it.

**Comment by jforberg** [comment]
@gaogaotiantian: I have made the requested changes; please review again

**Review by gaogaotiantian — APPROVED** [review]

**Comment by gaogaotiantian** [comment]
@ethanfurman is the component owner of tar related stuff. I'll leave the decision for him. The code looks good to me, but I'm not sure if there is any concern specifically for the optimization.

File: Lib/tarfile.py
Comment (picnixz) [inline Lib/tarfile.py]: I would split the comment (because gid -> gname is not the for this line, though one can understand). Maybe rename to `unames` and `gnames` since we only have `inodes` (it's not `inodes_cache`) or make them private (any reason to expose the attributes?)

File: Misc/NEWS.d/next/Library/2024-07-02-15-56-42.gh-issue-121267.yFBWkh.rst
Comment (picnixz) [inline Misc/NEWS.d/next/Library/2024-07-02-15-56-42.gh-issue-121267.yFBWkh.rst]:
```suggestion
Improve the performance of :mod:`tarfile` when writing files, by caching user names
```

**Comment by jforberg** [comment]
Thanks @picnixz, I have applied your suggestions.

**Comment by jforberg** [comment]
@ethanfurman, would you care to have a look at the patch? I think this performance gain would be quite beneficial for users of "tarfile".

**Comment by jforberg** [comment]
@picnixz @gaogaotiantian @ethanfurman My patch has been pending for a month or so and doesn't seem to be moving forward at the moment. Is there something further that I should be doing to help it along? Thanks for the help.

**Comment by picnixz** [comment]
> My patch has been pending for a month or so and doesn't seem to be moving forward at the moment. Is there something further that I should be doing to help it along

I personally have no way to commit. And patches may be pending for a long time until a core dev accepts it. I'm not a tarfile expert so I think you'll need to wait until Ethan has time to review it.

**Comment by jforberg** [comment]
@picnixz Thanks. I wasn't sure if people were waiting for me to do something at this point.

File: Lib/tarfile.py
Comment (hauntsaninja) [inline Lib/tarfile.py:2159]: It looks like the default value of tarinfo.uname/gname is the empty str, not None

Comment (jforberg) [reply Lib/tarfile.py:2159]: Thanks! Fixed

**Comment by morotti** [comment]
hello, not a core dev

i just noticed too that the performance of tarfile is horrible and found this ticket.

Commercial profiler run, using tarfile to tar the venv.
96 seconds to do the `tarfile.add()`, one third of that is looking up user/group info.
image

**Review by hauntsaninja — APPROVED** [review]
Thanks for the patch!

**Comment by ethanfurman** [comment]
Thanks everyone for moving this along and sorry I missed it -- my wife was in the hospital when this first started.

**Comment by jforberg** [comment]
Thanks for your help everyone! Ethan, I hope your wife is feeling better.

**Comment by picnixz** [comment]
@hauntsaninja  Should we backport the changes? (it may be nice and I don't think it breaks compatibility here)

> Ethan, I hope your wife is feeling better

I hope too!

**Comment by ethanfurman** [comment]
While it would be nice, it feels more like an enhancement and not a bug.

(And she is, thank you.)

**Comment by hauntsaninja** [comment]
We typically don't backport performance improvements. And hope things are well, Ethan!

**Comment by morotti** [comment]
it would be great to backport. it's a very small patch to make tarring twice as fast.

I wouldn't be surprised if this patch alone could save whole power-plant-worth-of-power from computers that are wasting time archiving stuff inefficiently. do we really have to wait for python 3.14 to land for the fix to be available?

**Comment by ethanfurman** [comment]
I've had enough "simple" patches with unintended consequences that I am not willing to backport this one.  If you would like to open a discourse thread about it and get consensus that this patch is fine to backport, I'll be happy to do so.

**Pull request merged**
